src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -250,29 +259,44 @@ object Ledger extends Logger {
     else Left("Transaction gas limit plus accumulated gas exceeds block gas limit")
   }
 
-  private def execute(stx: SignedTransaction, blockHeader: BlockHeader, worldStateProxy: InMemoryWorldStateProxy): Either[String, PR] = {
-    val programContext = ProgramContext[InMemoryWorldStateProxy, InMemoryWorldStateProxyStorage](stx, blockHeader, worldStateProxy)
-    val result = VM.run(programContext)
-    if (result.error.isDefined) Left(result.error.get.toString)
-    else {
-      if (stx.tx.isContractInit) payContractCreationCost(result).map(saveCreatedCost(programContext.env.ownerAddr, _))
-      else Right(result)
-    }
+  private def runVM(stx: SignedTransaction, blockHeader: BlockHeader, worldStateProxy: InMemoryWorldStateProxy): PR = {
+    val programContext: PC = ProgramContext(stx, blockHeader, worldStateProxy)
+    val result = vm.run(programContext)
+    if (stx.tx.isContractInit && result.error.isEmpty)
+      (payContractCreationCost _ andThen saveCreatedCode(programContext.env.ownerAddr))(result)
```

Comment:
Minor detail: In case `payContractCreationCost` results in an `OutOfGas` error, should we call `saveCreatedCode` either way?